### PR TITLE
feat: update IPC format to hold buffer_index

### LINKED ISSUE #786: ArrayView could cache cumulativenbuffers
In order to access any child of a arrayview we need to slice out it's buffers from the overall structure. Figuring out the buffer counts requires counting number of buffers in children. Since this value is accessed frequently we could give arrayview a slot to cache it

### PR DESCRIPTION
**Pull request by a10y**
Fixes #786 

Updates IPC format, removing the `Array.has_buffer` boolean field and replacing it with a `buffer_index`. The `buffer_index` for each array is an optional `uint64` that will hold the index into the `Arc<[Buffer]>` containing the buffer of the given array, or `null` if the array does not have a buffer.

Additional changes:

* Moves `children` into an `Arc<[Array]>` from `Vec<Array>` to improve cloning performance
* Uses `scalar_at_unchecked` in ChunkedArray::chunk to lookup chunk offsets, avoiding validity check since by construction, chunk offsets must be non-nullable

### CODE REVIEW

**Comment by a10y** [comment]
Seems like performance difference from the change is within margin of error for the benchmarks. Perhaps there are other places this can be used, like the `children` vector ArrayData, and correspondingly in VarBinArray (see https://github.com/apache/arrow-rs/issues/6408 for recent discussion about doing this in Arrow)

**Comment by robert3005** [comment]
ArrayData is the more common case. Views will be promoted to data pretty quickly if you perform operations on them

**Comment by robert3005** [comment]
~Fwiw if you want to try StructArray#project as well~ project is a bit special in how it selects things

File: vortex-serde/src/messages.rs [hunk 1]
```diff
@@ -18,7 +18,7 @@ pub enum IPCMessage<'a> {
 
 pub struct IPCSchema<'a>(pub &'a DType);
 pub struct IPCBatch<'a>(pub &'a Array);
-pub struct IPCArray<'a>(pub &'a Array);
+pub struct IPCArray<'a>(pub &'a Array, usize);
```
Comment (a10y) [inline vortex-serde/src/messages.rs:21]: the second argument here is the cumulative buffer index observed before array

**Comment by a10y** [comment]
The change to the IPC format to change `has_buffer` -> `buffer_index` seems to have improved pretty much all of the queries.

Q3, Q4, and Q21 in the PR benchmarks run actually got slower, but after multiple runs I'm unable to replicate that locally:


```
aduffy@DuffyProBook ~/c/vortex (aduffy/shared-vec)> critcmp before2 after2
group                              after2                                 before2
-----                              ------                                 -------
tpch_q1/vortex-file-compressed     1.00    464.9±8.26ms        ? ?/sec    1.03   477.6±16.36ms        ? ?/sec
tpch_q10/vortex-file-compressed    1.00    405.4±3.96ms        ? ?/sec    1.08    437.0±9.49ms        ? ?/sec
tpch_q11/vortex-file-compressed    1.00    132.1±9.72ms        ? ?/sec    1.12   147.6±12.85ms        ? ?/sec
tpch_q12/vortex-file-compressed    1.01   510.0±46.01ms        ? ?/sec    1.00   504.3±17.87ms        ? ?/sec
tpch_q13/vortex-file-compressed    1.00    136.9±3.33ms        ? ?/sec    1.08    147.7±6.12ms        ? ?/sec
tpch_q14/vortex-file-compressed    1.00     75.7±4.81ms        ? ?/sec    1.21     91.2±5.38ms        ? ?/sec
tpch_q15/vortex-file-compressed    1.00    130.5±4.44ms        ? ?/sec    1.38   179.5±28.35ms        ? ?/sec
tpch_q16/vortex-file-compressed    1.00     91.9±2.16ms        ? ?/sec    1.18   108.5±10.52ms        ? ?/sec
tpch_q17/vortex-file-compressed    1.00    325.4±5.26ms        ? ?/sec    1.17   382.0±35.44ms        ? ?/sec
tpch_q18/vortex-file-compressed    1.00   606.7±23.70ms        ? ?/sec    1.07   651.8±33.28ms        ? ?/sec
tpch_q19/vortex-file-compressed    1.00   612.6±43.75ms        ? ?/sec    1.08   664.5±43.12ms        ? ?/sec
tpch_q2/vortex-file-compressed     1.00     88.8±3.26ms        ? ?/sec    1.55   137.3±30.17ms        ? ?/sec
tpch_q20/vortex-file-compressed    1.00    216.2±9.73ms        ? ?/sec    1.03    223.1±7.02ms        ? ?/sec
tpch_q21/vortex-file-compressed    1.02  1163.8±47.95ms        ? ?/sec    1.00  1144.4±30.49ms        ? ?/sec
tpch_q22/vortex-file-compressed    1.00     80.7±2.34ms        ? ?/sec    1.03     83.5±1.55ms        ? ?/sec
tpch_q3/vortex-file-compressed     1.00    367.2±3.56ms        ? ?/sec    1.08   396.0±12.09ms        ? ?/sec
tpch_q4/vortex-file-compressed     1.00    316.2±4.71ms        ? ?/sec    1.06   335.2±10.87ms        ? ?/sec
tpch_q5/vortex-file-compressed     1.00    202.0±3.93ms        ? ?/sec    1.09    220.6±7.11ms        ? ?/sec
tpch_q6/vortex-file-compressed     1.00     68.8±2.04ms        ? ?/sec    1.05     72.4±4.83ms        ? ?/sec
tpch_q7/vortex-file-compressed     1.00    483.5±7.25ms        ? ?/sec    1.15   554.1±36.47ms        ? ?/sec
tpch_q8/vortex-file-compressed     1.00    211.7±3.07ms        ? ?/sec    1.06    223.4±3.39ms        ? ?/sec
tpch_q9/vortex-file-compressed     1.00    329.2±5.57ms        ? ?/sec    1.09   359.0±19.73ms        ? ?/sec
```

**Comment by robert3005** [comment]
Ship it. Didn’t know that we get cumulative nbuffers so often

**Comment by a10y** [comment]
I've updated the PR to remove `SharedVec`, since we were no longer slicing anything, instead opting to convert some of the `Vec<Buffer>` and `Vec<Array>` to `Arc<[T]>` equivalents

File: vortex-array/src/view.rs [hunk 2]
```diff
@@ -130,26 +130,13 @@ impl ArrayView {
...
-        let buffer_count = Self::cumulative_nbuffers(child);
-
         Ok(Self {
             encoding,
             dtype: dtype.clone(),
             len,
             flatbuffer: self.flatbuffer.clone(),
             flatbuffer_loc,
-            buffers: self.buffers[buffer_offset..][0..buffer_count].to_vec(),
+            buffers: self.buffers.clone(),
```
Comment (a10y) [inline vortex-array/src/view.rs:131]: note that we no longer need to slice, b/c each Array fb message contains the `buffer_index` field, which is a global offset into the buffers list

File: vortex-flatbuffers/src/generated/array.rs [hunk 3]
```diff
@@ -93,7 +93,7 @@ impl<'a> flatbuffers::Verifiable for Version {
 
 impl flatbuffers::SimpleToVerifyInSlice for Version {}
 pub enum ArrayOffset {}
-#[derive(Copy, Clone, PartialEq, Eq)]
+#[derive(Copy, Clone, PartialEq)]
```
Comment (a10y) [inline vortex-flatbuffers/src/generated/array.rs:96]: I'm not sure why the generator stripped `Eq` this time, but I don't think it matters

File: vortex-dtype/src/dtype.rs [hunk 4]
```diff
@@ -190,6 +190,7 @@ impl StructDType {
                 }
             };
 
+            // TODO: do this without the extra allocations.
```
Comment (a10y) [inline vortex-dtype/src/dtype.rs]: remove

File: vortex-array/src/view.rs [hunk 5]
```diff
@@ -173,20 +160,22 @@ impl ArrayView {
 
     /// Whether the current Array makes use of a buffer
     pub fn has_buffer(&self) -> bool {
-        self.flatbuffer().has_buffer()
+        self.flatbuffer().buffer_index().is_some()
     }
 
     /// The number of buffers used by the current Array and all its children.
     fn cumulative_nbuffers(array: fb::Array) -> usize {
```
Comment (robert3005) [inline vortex-array/src/view.rs]: This is no longer necessary?

Comment (robert3005) [reply vortex-array/src/view.rs]: Different signature

**Review by robert3005 — APPROVED** [review]

**Comment by a10y** [comment]
Hm, things seem to have gotten worse in the most recent benchmark run, despite it mainly just being deleting the SharedVec code. Several queries got 20-40% slower.

I can't replicate this level of slowdown locally

**Comment by robert3005** [comment]
As long as local run is faster I wouldn’t worry about one off runs. They're highly noisy

**Pull request merged**